components/calendar/src/chinese_based.rs

```diff
@@ -232,75 +414,159 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
+            );
+            let year = year_int.saturate();
+
+            let new_moon = Self::new_moon_before((date + 1).as_moment());
+            let month_i64 =
+                libm::round((new_moon - first_day_of_year) as f64 / MEAN_SYNODIC_MONTH) as i64 + 1;
+            debug_assert!(
+                ((u8::MIN as i64)..=(u8::MAX as i64)).contains(&month_i64),
+                "Month should be in range of u8! Value {month_i64} failed for RD {date:?}"
+            );
+            let month = month_i64 as u8;
+            let day_i64 = date - new_moon + 1;
+            debug_assert!(
+                ((u8::MIN as i64)..=(u8::MAX as i64)).contains(&month_i64),
+                "Day should be in range of u8! Value {month_i64} failed for RD {date:?}"
+            );
+            let day = day_i64 as u8;
+            let is_leap_year = Self::new_year_is_leap_year(
+                first_day_of_year,
+                Some(next_solstice),
+                Some(next_new_year),
+            );
+            let leap_month = if is_leap_year {
+                // This doesn't need to be checked for `None`, since `get_leap_month_from_new_year`
+                // will always return a number greater than or equal to 1, and less than 14.
+                NonZeroU8::new(Self::get_leap_month_from_new_year(first_day_of_year))
+            } else {
+                None
+            };
+            let cache = ChineseBasedCache {
+                new_year: first_day_of_year,
+                next_new_year,
+                leap_month,
+            };
+
+            // This can use `new_unchecked` because this function is only ever called from functions which
+            // generate the year, month, and day; therefore, there should never be a situation where
+            // creating this ArithmeticDate would fail, since the same algorithms used to generate the ymd
+            // are also used to check for valid ymd.
+            ChineseBasedDateInner(
+                ArithmeticDate::new_unchecked(year, month, day),
+                ChineseBasedYearInfo::Cache(cache),
+            )
+        }
     }
 
     /// Get a RataDie from a ChineseBasedDateInner
     ///
     /// This finds the RataDie of the new year of the year given, then finds the RataDie of the new moon
     /// (beginning of the month) of the month given, then adds the necessary number of days.
     pub(crate) fn fixed_from_chinese_based_date_inner(date: ChineseBasedDateInner<C>) -> RataDie {
-        let month = date.0.month as i64;
         let day = date.0.day as i64;
-        let first_day_of_year = date.1.new_year;
-        let month_approx = first_day_of_year + (month - 1) * 29;
-        let prior_new_moon = Self::new_moon_on_or_after(month_approx.as_moment());
+        let first_day_of_year = date.1.get_new_year();
+        let prior_new_moon = if let ChineseBasedYearInfo::Data(data) = date.1 {
+            first_day_of_year + data.last_day_of_previous_month(date.0.month).into()
```

Comment:
Nit: don't use `into` here. Not clear what you are going into. Without looking I assume  `new_moon_on_or_after` returns a `RataDie`.